tests/integration/test_sh_boot.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import os
+import subprocess
+import sys
+from typing import Text, Tuple
+
+import pytest
+
+from pex.testing import ALL_PY_VERSIONS, ensure_python_interpreter, run_pex_command
+from pex.typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from typing import Any, Iterable, Iterator, List
+
+
+@pytest.mark.parametrize(
+    ["execution_mode_args"],
+    [
+        pytest.param([], id="zipapp"),
+        pytest.param(["--venv"], id="venv"),
+    ],
+)
+def test_execute(
+    tmpdir,  # type: Any
+    execution_mode_args,  # type: List[str]
+):
+    # type: (...) -> None
+
+    cowsay = os.path.join(str(tmpdir), "cowsay.pex")
+    run_pex_command(
+        args=["cowsay==4.0", "-c", "cowsay", "-o", cowsay, "--bash-boot"] + execution_mode_args
+    ).assert_success()
+    assert "4.0" == subprocess.check_output(args=[cowsay, "--version"]).decode("utf-8").strip()
+
+
+def interpreters():
+    # type: () -> Iterable[Tuple[Text, List[Text]]]
+
+    def iter_interpreters():
+        # type: () -> Iterator[Tuple[Text, List[Text]]]
+
+        def entry(path):
+            # type: (Text) -> Tuple[Text, List[Text]]
+            return os.path.basename(path), [path]
+
+        yield entry(sys.executable)
+
+        for version in ALL_PY_VERSIONS:
+            interpreter = ensure_python_interpreter(version)
+            yield entry(interpreter)
+
+        locations = (
+            subprocess.check_output(
+                args=["/usr/bin/env", "bash", "-c", "command -v ash bash busybox dash ksh sh zsh"]
+            )
+            .decode("utf-8")
+            .splitlines()
+        )
+        for location in locations:
+            basename = os.path.basename(location)
+            if "busybox" == basename:
+                yield "ash (via busybox)", [location, "ash"]
+            else:
+                yield entry(location)
+
+    return sorted({name: args for name, args in iter_interpreters()}.items())
+
+
+@pytest.mark.parametrize(
+    ["interpreter_cmd"],
+    [pytest.param(args, id=name) for name, args in interpreters()],
+)
+def test_execute_via_interpreter(
+    tmpdir,  # type: Any
+    interpreter_cmd,  # type: List[str]
+):
+    # type: (...) -> None
+
+    cowsay = os.path.join(str(tmpdir), "cowsay.pex")
+    run_pex_command(
+        args=["cowsay==4.0", "-c", "cowsay", "-o", cowsay, "--bash-boot"]
+    ).assert_success()
+
+    assert (
+        "4.0"
+        == subprocess.check_output(args=interpreter_cmd + [cowsay, "--version"])
```

Comment:
The shell compatibility is good. On my machine I have everything except bare `ash` available:
```
$ tox -epy39-integration -q -- -vk test_execute_via_interpreter -n0 --no-header --no-summary
================================================================ test session starts ================================================================
collected 286 items / 276 deselected / 10 selected                                                                                                  

tests/integration/test_sh_boot.py::test_execute_via_interpreter[ash (via busybox)] PASSED                                                     [ 10%]
tests/integration/test_sh_boot.py::test_execute_via_interpreter[bash] PASSED                                                                  [ 20%]
tests/integration/test_sh_boot.py::test_execute_via_interpreter[dash] PASSED                                                                  [ 30%]
tests/integration/test_sh_boot.py::test_execute_via_interpreter[ksh] PASSED                                                                   [ 40%]
tests/integration/test_sh_boot.py::test_execute_via_interpreter[python] PASSED                                                                [ 50%]
tests/integration/test_sh_boot.py::test_execute_via_interpreter[python2.7] PASSED                                                             [ 60%]
tests/integration/test_sh_boot.py::test_execute_via_interpreter[python3.10] PASSED                                                            [ 70%]
tests/integration/test_sh_boot.py::test_execute_via_interpreter[python3.7] PASSED                                                             [ 80%]
tests/integration/test_sh_boot.py::test_execute_via_interpreter[sh] PASSED                                                                    [ 90%]
tests/integration/test_sh_boot.py::test_execute_via_interpreter[zsh] PASSED                                                                   [100%]

=================================================== 10 passed, 276 deselected, 1 warning in 8.68s ===================================================
______________________________________________________________________ summary ______________________________________________________________________
  py39-integration: commands succeeded
  congratulations :)
```